packages/fpdart/lib/src/effect.dart

```diff
@@ -0,0 +1,561 @@
+import 'dart:async';
+
+import './extension/future_or_extension.dart';
+import './extension/iterable_extension.dart';
+import 'exit.dart';
+import 'unit.dart' as fpdart_unit;
+
+part 'either.dart';
+part 'option.dart';
+
+typedef EffectGen<E, L> = ({
+  FutureOr<A> Function<A>(IEffect<E, L, A>) async,
+  A Function<A>(IEffect<E, L, A>) sync,
+});
+
+final class _EffectThrow<L> implements Exception {
+  final Cause<L> cause;
+  const _EffectThrow(this.cause);
+
+  @override
+  String toString() {
+    return "Effect.gen error: $cause";
+  }
+}
+
+EffectGen<E, L> _effectGen<E, L>(E env) => (
+      async: <A>(effect) => Future.sync(
+            () => effect.asEffect._unsafeRun(env).then(
+                  (exit) => switch (exit) {
+                    Left(value: final cause) => throw _EffectThrow<L>(cause),
+                    Right(value: final value) => value,
+                  },
+                ),
+          ),
+      sync: <A>(effect) {
+        final run = effect.asEffect._unsafeRun(env);
+        if (run is Future) {
+          throw _EffectThrow<L>(
+            Die.current(
+              Exception("gen.sync cannot execute async Effect"),
+            ),
+          );
+        }
+
+        return switch (run) {
+          Left(value: final cause) => throw _EffectThrow<L>(cause),
+          Right(value: final value) => value,
+        };
+      },
+    );
+
+typedef DoFunctionEffect<E, L, A> = FutureOr<A> Function(
+  EffectGen<E, L> $,
+);
+
+abstract interface class IEffect<E, L, R> {
+  const IEffect();
+  Effect<E, L, R> get asEffect;
+}
+
+final class Effect<E, L, R> extends IEffect<E, L, R> {
+  final FutureOr<Exit<L, R>> Function(E env) _unsafeRun;
+  const Effect._(this._unsafeRun);
+
+  @override
+  Effect<E, L, R> get asEffect => this;
+
+  @override
+  String toString() {
+    return "Effect(${_unsafeRun.runtimeType})";
+  }
+
+  /// {@category execution}
+  R runSync(E env) {
+    try {
+      final result = _unsafeRun(env);
+      if (result is Future) {
+        throw Die.current(
+          Exception("runSync cannot execute async Effect"),
+        );
+      }
+
+      return switch (result) {
+        Left(value: final cause) => throw cause,
+        Right(value: final value) => value,
+      };
+    } on Cause<L> {
+      rethrow;
+    } catch (error, stackTrace) {
+      throw Die(error, stackTrace);
+    }
+  }
+
+  /// {@category execution}
+  Exit<L, R> runSyncExit(E env) {
+    try {
+      final result = _unsafeRun(env);
+      if (result is Future) {
+        return Left(Die.current(
+          Exception("runSyncExit cannot execute async Effect"),
+        ));
+      }
+      return result;
+    } on Cause<L> catch (cause) {
+      return Left(cause);
+    } catch (error, stackTrace) {
+      return Left(Die(error, stackTrace));
+    }
+  }
+
+  /// {@category execution}
+  Future<R> runFuture(E env) async {
```

Comment:
Not sure if providing the env at the end is the best approach, as you might want to eliminate dependencies before running an Effect. One example is `Scope`, if you plan to add that eventually.

Also one big missing piece here is async dependency injection. I attempted to solve this with a simple Layer/Runtime implementation in elemental, but maybe you can come up with something else.

Reply:
I added `provide` and `provideEffect` to provide the full `E` before running:
https://github.com/SandroMaglione/fpdart/blob/a3352b5de8e1c9512ce7abb66834c500fc072f10/packages/fpdart/lib/src/effect.dart#L293-L304

`provideEffect` works for "async dependency", or are you meaning something else?

***

The problem in dart is that you cannot erase part of the dependency, so you must provide all `E`.

Do you have any idea for some way to solve this issue?

Reply:
No you can't really get around that in dart. Which is why Effect's solution for dependencies isn't really a nice option here.